orttraining/orttraining/training_ops/cuda/communication/recv.cc

```diff
@@ -3,75 +3,112 @@
 
 #ifdef USE_HOROVOD
 
-#include "recv.h"
+#include "orttraining/training_ops/cuda/communication/recv.h"
+#include "orttraining/training_ops/cuda/communication/common.h"
+#include <mpi.h>
 
 namespace onnxruntime {
 namespace cuda {
 
-#define REGISTER_RECV_KERNEL_TYPED(T)                                    \
-  ONNX_OPERATOR_TYPED_KERNEL_EX(                                         \
-      Recv,                                                              \
-      kMSDomain,                                                         \
-      1,                                                                 \
-      T,                                                                 \
-      kCudaExecutionProvider,                                            \
-      KernelDefBuilder()                                                 \
-          .InputMemoryType<OrtMemTypeCPUInput>(0)  /* CPU variable */    \
-          .OutputMemoryType<OrtMemTypeCPUOutput>(0)  /* CPU variable */  \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())         \
-          .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>()), \
-      Recv<T>);
-
-REGISTER_RECV_KERNEL_TYPED(MLFloat16)
-REGISTER_RECV_KERNEL_TYPED(float)
-REGISTER_RECV_KERNEL_TYPED(double)
+ONNX_OPERATOR_KERNEL_EX(
+    Recv,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
+        .InputMemoryType<OrtMemTypeCPUInput>(1)   /* CPU variable */
+        .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>())
+        .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
```

Comment:
Does FixedSize mean all tensor shapes should be known before computation?

Reply:
FixedSize means bitwidth is fixed, aka everything except string.